feat: extend functionality of `Workspace` class

alvarobartt opened the pull request:
# Description

This PR extends the functionality of the `Workspace` class created as of Argilla 1.8.0, to not just do name-to-id, but also to be able to manage the workspaces and the users assigned to those, as the API supports those features via the `api/workspaces` endpoints.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add unit tests

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
fix: remove unused `AuthenticatedClient` import

alvarobartt committed:
feat: add sdk-support for `api/v1/workspaces`

alvarobartt committed:
fix: `WorkspaceModel.id` to be UUID instead of `str`

alvarobartt committed:
feat: add temporary `WorkspaceUserModel`

To be removed and replaced with the `UserModel` created upon https://github.com/argilla-io/argilla/pull/3169 completion

alvarobartt committed:
feat: add sdk-support for `/api/workspaces`

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/workspace-management

alvarobartt committed:
feat: use `construct` to skip validation

alvarobartt committed:
feat: add `Workspace` class to manage workspaces

alvarobartt committed:
fix: exception handling in `Workspace`

alvarobartt committed:
test: add unit tests for `Workspace`

alvarobartt committed:
test: add unit tests for `/api/workspaces` in SDK

alvarobartt committed:
feat: remove to-be-deprecated `.from_name` calls

alvarobartt committed:
docs: update `CHANGELOG.md`

frascuchon commented on src/argilla/client/feedback/utils.py:
```diff
@@ -111,10 +111,10 @@ def feedback_dataset_in_argilla(
     if (name and workspace) or name:
         if workspace is None:
-            workspace = rg.Workspace.from_name(rg.active_client().get_workspace())
+            workspace = rg.Workspace(rg.active_client().get_workspace())
```
That means that `Workspace.__init___` will call internally the argilla API, right? I would prefer to have this operation outside of an object constructor. I would say to separate the `Workspace` class for the data model and `Workspaces` class to interact with workspaces in Argilla.
```python
import argilla as rg 

workspace= rg.Workspaces.from_name("my-workspace")
workspace.id # UUID(......)
workspace.name # my-workspace

new_workspace= rg.Workspaces.create(Workspace(name="new-workspace")
new_workspace.id # UUID (...)
new_workspace.name # new-workspace
```

This approach allows us to decouple the internal API integration (will be in Workspaces) from the workspace model itself.

alvarobartt replied:
This is what we had before, but it's a little bit weird because users will be able to register workspaces that don't exist in Argilla, and this functionality is just intended to be used within Argilla, but if you think it's better to keep the init simpler then we can just do nothing on init, and keep the `from_name` which is still there

frascuchon commented on src/argilla/client/sdk/v1/workspaces/api.py:
```diff
@@ -0,0 +1,54 @@
+
+def get_workspace(
+    client: httpx.Client,
+    id: str,
```
We can pass a UUID since workspace ids will be always an UUID 
```suggestion
    id: UUID,
```

frascuchon commented on src/argilla/client/sdk/workspaces/api.py:
```diff
@@ -42,7 +40,121 @@ def list_workspaces(client: httpx.Client) -> Response[Union[List[WorkspaceModel]
 
     if response.status_code == 200:
-        parsed_response = [WorkspaceModel(**workspace) for workspace in response.json()]
+        parsed_response = [WorkspaceModel.construct(**workspace) for workspace in response.json()]
```
By using the `construct` method, all automatic casting will be applied. For example `str`-> `UUID` or `str`-> `datetime` will be applied?

alvarobartt replied:
No it wont', so maybe we just prefer to avoid `construct` usage? Even though it will just affect the `id` and the `inserted/created_at` attributes (and the last two ones are not to be modified, and a `str` formatted as an UUID can be parsed as UUID, so I see no issue. But if we want to be more strict with typing sure, we can just just avoid the `construct` to validate the fields (I was just using the latter when there were nested `pydantic.BaseModel`s)

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
+    inserted_at: datetime
+    updated_at: datetime
+
+    def __init__(self, name: Optional[str] = None, *, id: Optional[str] = None) -> None:
```
```suggestion
    def __init__(self, name: Optional[str] = None, *, id: Optional[Union[str,UUID]] = None) -> None:
```

but in any case we should check that id contains a valid UUID

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
+                    "create it via the `Workspace.create` method as follows:"
+                    f" `Workspace(name=`{name}`)`."
+                )
+        self.__dict__.update(existing_workspace.dict())
```
What this line exactly does?

alvarobartt replied:
This line is just mapping the `WorkspaceModel` to the current model, so all the attributes are mapped, instead of having to manually set all the attributes one by one e.g. `self.name = existing_workspace.name`

alvarobartt replied:
That's also why I've defined the attributes on the top of the class before the `__init__`

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
+            A list of `WorkspaceUserModel` instances.
+        """
+        if not hasattr(self, "__users") or self.__users is None:
+            self.__users = workspaces_api.list_workspace_users(active_client().http_client.httpx, self.id).parsed
```
We should reduce the couple with the active_client logic. I think the httpx client should be part of the workspace on object initialization. Then we can do:
```suggestion
            self.__users = workspaces_api.list_workspace_users(self.__httpx, self.id).parsed
```
otherwise, we can start having problems in multi-thread multi client environments (a process connecting to different argilla server to sync some datasets/workspaces)

alvarobartt replied:
Sure, then I'll also add the client on top to be reused in the whole class :smile:

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
-        self.id = id or None
+        try:
+            created_user = workspaces_api.create_workspace_user(
+                client=active_client().http_client.httpx,
```
same here
```suggestion
                client=self.__httpx,
```

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
+    def __str__(self) -> str:
+        return f"Workspace(id={self.id}, name={self.name}, users={self.users}, insterted_at={self.inserted_at}, updated_at={self.updated_at}"
+
+    def add_user(self, id: str) -> None:
```
put UUID instead of str

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -14,67 +13,238 @@
-        self.name = name
-        self.id = id or None
+        try:
+            created_user = workspaces_api.create_workspace_user(
```
you're creating a workspace-user relationship, right?

alvarobartt replied:
Yes, I just renamed the method to be `add_user` in the `Workspace` class just to avoid using the `create_workspace_user` naming as I found it a little bit confusing, while `add_user` seems more straight forward

frascuchon commented on tests/client/test_workspaces.py:
```diff
@@ -17,16 +17,121 @@
+def test_workspace_from_name(mocked_client: SecuredClient) -> None:
+    the_api = api.active_api()
+    workspace = the_api.http_client.post("/api/workspaces", json={"name": "test_workspace"})
+    assert workspace["name"] == "test_workspace"
```
You can setup your test by using the DB factories.
```suggestion
    workspace = WorkspaceFactory.create(name="test_workspace") # this will create a workspace in DB that is accessible through the API
```

alvarobartt replied:
Right, I'll refactor the unit tests then!

frascuchon commented on tests/client/test_workspaces.py:
```diff
@@ -17,16 +17,121 @@
+def test_workspace_init(mocked_client: SecuredClient) -> None:
+    the_api = api.active_api()
+    workspace = the_api.http_client.post("/api/workspaces", json={"name": "test_workspace"})
+    assert workspace["name"] == "test_workspace"
```
same here, use test factories instead, will simplify your test

frascuchon commented on tests/client/test_workspaces.py:
```diff
@@ -17,16 +17,121 @@
+        },
+    )
+    assert user["username"] == "test_user"
+    assert isinstance(user["id"], str)
```
Again you can use the UserFactory to create a user fo the test

alvarobartt committed:
test: use pre-defined factories instead

Co-authored-by: Francisco Aranda <francis@argilla.io>

gabrielmbmb commented on src/argilla/client/sdk/workspaces/models.py:
```diff
@@ -14,12 +13,33 @@
+from typing import List, Optional
+from uuid import UUID
+
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal
```
I think that we can just import `from typing import Literal` without trying to catch the exception. `Literal` was added in Python 3.8, which is the oldest version that we support, right?

I think that I've seen this in some other places, we should update it everywhere.

alvarobartt replied:
So we've already removed support for Python 3.7? I can create a separate PR to remove the unsupported stuff from now on 👍🏻

alvarobartt committed:
fix: share `httpx.Client` & detach API calls from `__init__`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
fix: apply `UUID` conversion if `id` is str

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
revert: keep `from_name` when retrieving `Workspace`

alvarobartt committed:
docs: update docstrings to point to `Workspace.from_name`

alvarobartt committed:
test: update `Workspace` unit tests

alvarobartt committed:
feat: clean `workspaces` sdk v0 and v1

- remove unused `.format`
- expect `UUID` instead of `str` for IDs
- don't use `construct` to avoid losing types

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,281 @@
+            ...     updated_at=datetime.datetime(2021, 8, 31, 10, 0, 0),
+            ... )
         """
+        self.client = client
```
I think we can keep this attribute private, since won't be used outside this class
```suggestion
        self._client = client
```

alvarobartt replied:
Ok I'll make it private then! 👍🏻

alvarobartt committed:
test: fix assertion on `UUID`

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,281 @@
+            A list of `WorkspaceUserModel` instances.
+        """
+        if not hasattr(self, "__users") or self.__users is None:
+            self.__users = workspaces_api.list_workspace_users(self.client, self.id).parsed
```
Should be better to create the __users attribute on init, just to know what attributes (public and privates) contain the instances by reading the init method?

alvarobartt replied:
Since this is just intended to be used internally, we're just defining `users` as that's the property the users will use to list the linked users to a given workspace, there's no need to create `self.__users`, as it won't be accessed/used until `self.users` is accessed.

alvarobartt replied:
Since this is just intended to be used internally, we're just defining `users` as that's the property the users will use to list the linked users to a given workspace, there's no need to create `self.__users`, as it won't be accessed/used until `self.users` is accessed.

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,281 @@
+            >>> workspace.add_user("my-user-id")
+        """
+        try:
+            created_user = workspaces_api.create_workspace_user(
```
should we check if the user is found in the local store before?

alvarobartt replied:
This is what we had before, a validation layer against Argilla in the `__init__`, but since we decided to remove it, I think we can just rely on the API validation included as part of those endpoints

alvarobartt replied:
Either calling the API to check whether the user exists or not and then to add it to the workspace (2 API calls) or just add it to the workspace (1 API call) and if it fails we handle the exception to show either that the user doesn't exist and/or the user is already there, in case it fails.

alvarobartt committed:
feat: add `Workspace.list` classmethod

alvarobartt committed:
test: add `Workspace.list` unit tests

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,281 @@
+            >>> workspace.delete_user("my-user-id")
+        """
+        try:
+            workspaces_api.delete_workspace_user(
```
same here, maybe we need to find it in list before, just for early validations

alvarobartt replied:
So same response as above, is it worth to have another API call to repeat a validation that is already included as part of the endpoints we're calling? Isn't just better to catch the exceptions and show the relevant error messages based on those?

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,303 @@
+        else:
+            self.__users.append(created_user)
+
+    # TODO(alvarobartt): also allow user addition via name (which is more user-friendly)
```
Regarding this, once we have the `User` classes we can pass User instances instead of IDs or names and let the responsibility of find users by name to the `User.from_name` class method

alvarobartt replied:
Yes, indeed I'm already preparing the user-management layer so that `users` contains a list of validated users instead, but I agree that we can either allow id, name, and or `User` instance

alvarobartt committed:
fix: make `Workspace.client` private (`__client`)

Co-authored-by: Francisco Aranda <francis@argilla.io>

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,303 @@
+        )
+
+    # TODO(alvarobartt): also allow user addition via name (which is more user-friendly)
+    def add_user(self, id: str) -> None:
```
I'm missing a way to refresh o resync the workspace with Argilla server. If an external process is adding or deleting new users, it would be nice to refresh data locally to fetch latest changes without a new `Workspace.from_id(ws.id)`

A workflow could be:
```python
workspace = Workspace.from_name("ws")
workspace.users # []
... # here another process added 2 users (user_id=1 and user_id=2) to this workspace
workspace.add_user(user_id=1) # this will fail because the user already exists in this workspace
workspace.refresh()
workspace.users # will return the updated user list
```

alvarobartt replied:
Sure I can easily included, I just didn't think about it as a possible workflow, but you never know 😆

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,303 @@
+        )
+
+    # TODO(alvarobartt): also allow user addition via name (which is more user-friendly)
+    def add_user(self, id: str) -> None:
```
we can explicit the id is for a user
```suggestion
    def add_user(self, user_id: str) -> None:
```

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,303 @@
+            self.__users.append(created_user)
+
+    # TODO(alvarobartt): also allow user addition via name (which is more user-friendly)
+    def delete_user(self, id: str) -> None:
```
same here
```suggestion
    def delete_user(self, user_id: str) -> None:
```

alvarobartt committed:
fix: `users` calls API always & rename `id`->`user_id`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/workspace-management

alvarobartt committed:
fix: enforce `UUID` instead of `str`

As suggested by @frascuchon, we should enforce the IDs to be UUIDs instead of strings, so we can no longer use `.construct`

alvarobartt committed:
fix: call `rg.active_client` once in `feedback_dataset_in_argilla`

alvarobartt committed:
fix: `FeedbackDatasetModel` outdated `datetime` fields

alvarobartt committed:
fix: use f-string instead of `format`

alvarobartt committed:
fix: init `Workspace` only via classmethods

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

alvarobartt committed:
test: handle exceptions in `Workspace.__init__`

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,305 @@
+        try:
+            ws = workspaces_api.create_workspace(client=instance.__client, name=name).parsed
+            instance.__dict__.update(ws.dict())
+            return instance
```
Maybe we can create a new instance after we're sure everything went well. 
```suggestion
        try:
            ws = workspaces_api.create_workspace(client=instance.__client, name=name).parsed
            # This code block is a potential WorspaceFactory method
            instance = cls.__new__(cls)
            instance.__client = cls.__active_client()
            instance.__dict__.update(ws.dict())
            return instance
```

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,305 @@
+        try:
+            ws = workspaces_api_v1.get_workspace(instance.__client, id).parsed
+            instance.__dict__.update(ws.dict())
+            return instance
```
same here, and I think we can reuse the new-instance code block

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,305 @@
+        for ws in workspaces:
+            if ws.name == name:
+                instance.__dict__.update(ws.dict())
+                return instance
```
same here, I think we can first find the proper workspace and then create the instance with workspace + client

frascuchon commented on src/argilla/client/workspaces.py:
```diff
@@ -13,68 +12,305 @@
+                instance = cls.__new__(cls)
+                instance.__client = client
+                instance.__dict__.update(ws.dict())
+                yield instance
```
code duplication

alvarobartt committed:
feat: add private classmethod `__new_instance`

alvarobartt committed:
Merge branch 'develop' into feat/workspace-management

alvarobartt committed:
Merge branch 'develop' into feat/workspace-management
